File: Prototype/BigGp/src/Model/SmallNodes/Expressions/Logic/Childs/BooleanExpression.java
package Model.SmallNodes.Expressions.Logic.Childs;

import Model.Interfaces.Crossable;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Expressions.Logic.LogicExpression;
import Model.SmallNodes.Operators.Logic.BoolOperator;
import Serialization.Token;

import java.util.ArrayList;

public class BooleanExpression extends Node implements Crossable, SubtreeMutable {

    private LogicExpression leftExpression;
    private BoolOperator operator;
    private LogicExpression rightExpression;

    @Override
    public String toString() {
        return "(" + leftExpression.toString() + operator.toString() + rightExpression.toString() + ")";
    }

    public void setLogicExpression(Node oldExpr, LogicExpression logicExpression) {
        if (leftExpression.toString().equals(oldExpr.toString()) && leftExpression.hashCode() == oldExpr.hashCode()) {
            this.leftExpression = logicExpression;
        } else {
            this.rightExpression = logicExpression;
        }
    }
    @Override
    public void subtreeMutate() {
        if (this.getParentNode() instanceof LogicExpression logicExpression) {
            logicExpression.setBooleanExpression(new BooleanExpression(logicExpression));
        }
    }

    public ArrayList<Node> getChildrenAsNodes(){
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);

        nodes.addAll(leftExpression.getChildrenAsNodes());
        nodes.addAll(operator.getChildrenAsNodes());
        nodes.addAll(rightExpression.getChildrenAsNodes());

        return nodes;
    }

    public BooleanExpression(Node parentNode) {
        super(parentNode, "BooleanExpression", true);

        double lastNextLogicExpressionChance = ((LogicExpression) parentNode).getNEXT_LOGIC_BOOLEAN_EXPRESSION_CHANCE();
        leftExpression = new LogicExpression(this, lastNextLogicExpressionChance);
        operator = new BoolOperator(this);
        rightExpression = new LogicExpression(this, lastNextLogicExpressionChance);
    }

    public BooleanExpression(Node parentNode, ArrayList<Token> tokens) {
        super(parentNode, "BooleanExpression", true);
        tokens.remove(0);

        double lastNextLogicExpressionChance = ((LogicExpression) parentNode).getNEXT_LOGIC_BOOLEAN_EXPRESSION_CHANCE();
        leftExpression = new LogicExpression(this, lastNextLogicExpressionChance, tokens);
        operator = new BoolOperator(this, tokens);
        rightExpression = new LogicExpression(this, lastNextLogicExpressionChance, tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Small/Expressions.Logic/BooleanExpression.cs
using Model.Interfaces.Evolution;
using Model.Nodes.Small.Operators.Logic;

namespace Model.Nodes.Small.Expressions.Logic;

public class BooleanExpression : Node, ICrossable, ISubtreeMutable
{
    private LogicExpression Left { get; set; }
    private BoolOperator Operator { get; set; }
    private LogicExpression Right { get; set; }

    public void SetLogicExpression(Node oldExpression, LogicExpression newExpression)
    {
        if (Left == oldExpression)
        {
            Left = newExpression;
        }
        else if (Right == oldExpression)
        {
            Right = newExpression;
        }
        else
        {
            throw new InvalidOperationException("Expression not found");
        }
    }

    public override List<Node> ChildrenAsNodes()
    {
        var nodes = new List<Node> { this };
        nodes.AddRange(Left.ChildrenAsNodes());
        nodes.AddRange(Operator.ChildrenAsNodes());
        nodes.AddRange(Right.ChildrenAsNodes());

        return nodes;
    }

    public void SubtreeMutate()
    {
        if (ParentNode is LogicExpression expression)
        {
            expression.BooleanExpression = new BooleanExpression(expression);
        }
    }

    public BooleanExpression(Node parentNode) : base(parentNode, "BooleanExpression", true)
    {
        var lastNextLogicExpressionChance = ParentNode is LogicExpression logicExpression
            ? logicExpression.NextLogicBooleanExpressionChance
            : throw new InvalidOperationException("Parent node is not a logic expression");

        Left = new LogicExpression(this, lastNextLogicExpressionChance);
        Operator = new BoolOperator(this);
        Right = new LogicExpression(this, lastNextLogicExpressionChance);
    }

    public BooleanExpression(Node parentNode, List<Token> tokens) : base(parentNode, "BooleanExpression", true)
    {
        tokens.RemoveAt(0);

        var lastNextLogicExpressionChance = ParentNode is LogicExpression logicExpression
            ? logicExpression.NextLogicBooleanExpressionChance
            : throw new InvalidOperationException("Parent node is not a logic expression");

        Left = new LogicExpression(this, tokens, lastNextLogicExpressionChance);
        Operator = new BoolOperator(this, tokens);
        Right = new LogicExpression(this, tokens, lastNextLogicExpressionChance);
    }

    public override string ToString() => $"({Left} {Operator} {Right})";
}